File: Advanced Robotics Course/Chapter11/Lesson3/Chapter11_Lesson3.cpp
#include <vector>
#include <Eigen/Dense>

using Eigen::VectorXd;
using Eigen::MatrixXd;

struct Transition {
    int s;
    int a;
};

struct Demonstration {
    std::vector<Transition> traj;
};

VectorXd phi(int s, int a, int k);

// Empirical discounted feature expectation from demonstrations
VectorXd empiricalFeatureExpectation(const std::vector<Demonstration>& demos,
                                     double gamma,
                                     int k) {
    VectorXd mu_E = VectorXd::Zero(k);
    for (const auto& demo : demos) {
        double discount = 1.0;
        for (std::size_t t = 0; t < demo.traj.size(); ++t) {
            int s = demo.traj[t].s;
            int a = demo.traj[t].a;
            mu_E += discount * phi(s, a, k);
            discount *= gamma;
        }
    }
    mu_E /= static_cast<double>(demos.size());
    return mu_E;
}

// Simple gradient update for theta in linear reward R_theta(s,a) = theta^T phi(s,a)
void gradientUpdate(VectorXd& theta,
                    const VectorXd& mu_E,
                    const VectorXd& mu_theta,
                    double alpha) {
    VectorXd grad = mu_E - mu_theta;
    theta += alpha * grad;
}
      

File: Tutorials/Codes/Advanced_Robotics_Course/Chapter11/Lesson3/Chapter11_Lesson3.java
import java.util.List;

public class LinearIrl {

    // phi(s,a) is assumed known and returns double[k]
    public static double[] phi(int s, int a, int k) {
        double[] f = new double[k];
        // TODO: fill with task-specific features
        return f;
    }

    public static double[] empiricalFeatureExpectation(List<int[][]> demos,
                                                       double gamma, int k) {
        double[] muE = new double[k];
        for (int[][] traj : demos) {
            double discount = 1.0;
            for (int t = 0; t < traj.length; t++) {
                int s = traj[t][0];
                int a = traj[t][1];
                double[] f = phi(s, a, k);
                for (int i = 0; i < k; i++) {
                    muE[i] += discount * f[i];
                }
                discount *= gamma;
            }
        }
        int n = demos.size();
        for (int i = 0; i < k; i++) {
            muE[i] /= (double) n;
        }
        return muE;
    }

    public static void gradientUpdate(double[] theta,
                                      double[] muE,
                                      double[] muTheta,
                                      double alpha) {
        for (int i = 0; i < theta.length; i++) {
            double grad = muE[i] - muTheta[i];
            theta[i] += alpha * grad;
        }
    }
}
      
